src/mono/wasm/runtime/invoke-js.ts

```diff
@@ -0,0 +1,316 @@
+import { Int32Ptr } from "./types/emscripten";
+
+const cs_to_js_map = new Map<MonoType, MarshallerToJs>();
+const js_to_cs_map = new Map<MonoType, MarshallerToCs>();
+const _invoke_js_function_cache = new Map<MonoString, Function>(); // this is never collected
+
+/**
+ * Layout of the buffers would be
+ * signature: [TRes,T1,T2,T3]
+ *   - TRes === MonoTypeNull means void function
+ * data: [res*,t1*,t2*,t3*]
+ */
+
+// C# signature would be `static TRes Invoke<TRes>()` or `static void Invoke()`
+export function mono_wasm_invoke_js_0(interned_function_id: MonoString, signature: VoidPtr, data: VoidPtr): InvokeJSResult {
+    try {
+        const fn = mono_wasm_lookup_function(interned_function_id);
+        const res_js = fn();
+        covert_result_to_cs(signature, data, res_js);
+        return InvokeJSResult.Success;
+    } catch (ex) {
+        _marshal_exception_to_cs(data, ex);
+        return InvokeJSResult.Exception;
+    }
+}
+
+// C# signature would be `static TRes Invoke<T1, TRes>(ref T1 t1)` or `static void Invoke<T1>(ref T1 t1)`
+export function mono_wasm_invoke_js_1(interned_function_id: MonoString, signature: VoidPtr, data: VoidPtr): InvokeJSResult {
+    //console.log("mono_wasm_invoke_js_1 A " + signature + " " + data);
+    //console.log("mono_wasm_invoke_js_1 S0 " + getU32(<any>signature + (0 << 2)));
+    //console.log("mono_wasm_invoke_js_1 S1 " + getU32(<any>signature + (1 << 2)));
+    //console.log("mono_wasm_invoke_js_1 D0 " + getU32(<any>data + (0 << 2)));
+    //console.log("mono_wasm_invoke_js_1 D1 " + getU32(<any>data + (1 << 2)));
+    try {
+        const fn = mono_wasm_lookup_function(interned_function_id);
+        //console.log("mono_wasm_invoke_js_1 B");
+        const arg1_js = covert_arg_to_js(signature, data, 1);
+        //console.log("mono_wasm_invoke_js_1 C");
+        const res_js = fn(arg1_js);
+        //console.log("mono_wasm_invoke_js_1 D");
+        covert_result_to_cs(signature, data, res_js);
+        //console.log("mono_wasm_invoke_js_1 E");
+        return InvokeJSResult.Success;
+    } catch (ex) {
+        //console.log("mono_wasm_invoke_js_1 F");
+        _marshal_exception_to_cs(data, ex);
+        return InvokeJSResult.Exception;
+    }
+}
+
+export function mono_wasm_bind_js_function(interned_function_id: MonoString, arg_count: number, signature: VoidPtr, function_js_handle: Int32Ptr, is_exception: Int32Ptr): MonoObject {
+    try {
+        const fn = mono_wasm_lookup_function(interned_function_id);
+        const closure: any = { MONO, BINDING, INTERNAL, fn };
+        let body = "try {\n";
+
+        for (let position = 1; position <= arg_count; position++) {
+            const offset = position << 2;
+            const mono_type = <MonoType><any>getU32(<any>signature + offset);
+            const covertor = cs_to_js_map.get(mono_type);
+            assert(covertor, `Unknow covertor for type ${mono_type} at ${position}`);
```

Comment:
Not remotely a problem while it's a draft, but we'll need to spell-check all the English in user-facing stuff like error messages, argument names, etc.

Reply:
Sure, we are not even close to end of prototyping here.